exchanges/account/account.go

```diff
@@ -71,48 +81,210 @@ func Process(h *Holdings) error {
+				balance.m.Lock()
+				currencyBalances[target] = Balance{
+					CurrencyName:           currency.Code{Item: item, UpperCase: true},
+					Total:                  balance.total,
+					Hold:                   balance.hold,
+					Free:                   balance.free,
+					AvailableWithoutBorrow: balance.availableWithoutBorrow,
+					Borrowed:               balance.borrowed,
+				}
+				balance.m.Unlock()
+				target++
+			}
+
+			if len(currencyBalances) == 0 {
+				continue
+			}
+
+			accountsHoldings = append(accountsHoldings, SubAccount{
+				ID:         subAccount,
+				AssetType:  ai,
+				Currencies: currencyBalances,
+			})
+			break // Don't continue to iterate through the other assets.
 		}
 	}
-	return Holdings{}, fmt.Errorf("%v holdings data not found for %s", assetType, exch)
+
+	if len(accountsHoldings) == 0 {
+		return Holdings{}, fmt.Errorf("%s %s %w",
+			exch,
+			assetType,
+			errAssetHoldingsNotFound)
+	}
+	return Holdings{Exchange: exch, Accounts: accountsHoldings}, nil
+}
+
+// GetBalance returns the internal balance for that asset item.
+func GetBalance(exch, subAccount string, ai asset.Item, c currency.Code) (*BalanceInternal, error) {
+	if exch == "" {
+		return nil, errExchangeNameUnset
+	}
+
+	if !ai.IsValid() {
+		return nil, fmt.Errorf("%s %w", ai, asset.ErrNotSupported)
+	}
+
+	if c.IsEmpty() {
+		return nil, currency.ErrCurrencyCodeEmpty
+	}
+
+	exch = strings.ToLower(exch)
+	subAccount = strings.ToLower(subAccount)
+
+	service.m.Lock()
+	defer service.m.Unlock()
+
+	accounts, ok := service.exchangeAccounts[exch]
+	if !ok {
+		return nil, errExchangeHoldingsNotFound
+	}
+
+	assetBalances, ok := accounts.SubAccounts[subAccount]
+	if !ok {
+		return nil, errNoExchangeSubAccountBalances
+	}
+
+	currencyBalances, ok := assetBalances[ai]
+	if !ok {
+		return nil, errAssetHoldingsNotFound
+	}
+
+	bal, ok := currencyBalances[c.Item]
+	if !ok {
+		return nil, errNoBalanceFound
+	}
+	return bal, nil
 }
 
 // Update updates holdings with new account info
 func (s *Service) Update(a *Holdings) error {
 	exch := strings.ToLower(a.Exchange)
-	s.Lock()
-	acc, ok := s.accounts[exch]
+	s.m.Lock()
+	defer s.m.Unlock()
+	accounts, ok := s.exchangeAccounts[exch]
 	if !ok {
 		id, err := s.mux.GetID()
 		if err != nil {
-			s.Unlock()
 			return err
 		}
+		accounts = &Accounts{
+			ID:          id,
+			SubAccounts: make(map[string]map[asset.Item]map[*currency.Item]*BalanceInternal),
+		}
+		s.exchangeAccounts[exch] = accounts
+	}
+
+	for x := range a.Accounts {
+		lowerSA := strings.ToLower(a.Accounts[x].ID)
+		var accountAssets map[asset.Item]map[*currency.Item]*BalanceInternal
+		accountAssets, ok = accounts.SubAccounts[lowerSA]
+		if !ok {
+			accountAssets = make(map[asset.Item]map[*currency.Item]*BalanceInternal)
+			accounts.SubAccounts[lowerSA] = accountAssets
+		}
+
+		if !a.Accounts[x].AssetType.IsValid() {
+			return fmt.Errorf("cannot load sub account holdings for %s [%s] %w",
```

Comment:
If one asset type is incorrect, should all account holdings be dismissed? What if the problem asset is the second in the list? The first balance has been loaded in your new `BalanceInternal` on line 233 and you will have a discrepancy between balance internal and elsewhere (since it gets updated, but not published). One of those things needs to change